### ISSUE TITLE
wesnothd and campaignd leak connections

### ISSUE THREAD

**Issue by Pentarctagon** [wesnoth/wesnoth#6486]
Currently the official wesnothd and campaignd instances slowly leak connections over time, resulting in them eventually becoming unresponsive due to hitting the system limit of the maximum number of files (read: sockets) open. This logs an error such as:
`error server: Accept failed: Too many open files`
Then when looking at it with `lsof`, will show connections of type `sock`, which as per its man page means "a socket of unknown domain".

This can also be seen by viewing the graph of connections to wesnoth [here](http://wesnothd.wesnoth.org/) when looking at timescales of a month or more - the graph trends upwards even though the number of players isn't meaningfully changing in the same time period, and resets once wesnothd is restarted.

---

The current suspicion at the moment is that there's a `boost::coroutines::detail::forced_unwind` exception being thrown somewhere and handled by our own code, which is incorrect since that should instead be allowed to propagate back up to the boost coroutine code.

For example, in `wesnothd/game.cpp`'s destructor, there are things that could call a `forced_unwind` exception, but it's wrapped in a try/catch(...) block. However we're not sure if this particular example is actually causing the problem for wesnothd, and even if it was it wouldn't account for the same issue happening in campaignd as well.

**Comment by Pentarctagon** [comment]
@loonycyborg FYI

**Comment by loonycyborg** [comment]
Seems the link to wesnothd.wesnoth.org in OP doesn't work.

**Comment by Pentarctagon** [comment]
Weird, should be fixed now.

**Comment by Pentarctagon** [comment]
Some further updates:

I was able to reproduce getting a sock line to show up in the lsof output locally for wesnothd. steps to reproduce:
* Put a breakpoint at the start of the game destructor.
* Host a game.
* Have all players leave the game, thus ending it.
* After the breakpoint triggers, have the players also disconnect from the server entirely.
* Let wesnothd continue past the breakpoint.

It doesn't reach the forced_rewind catch block I added, but does log:
```
20220205 17:55:04 info server: 127.0.0.1        p2      ended game:     "p2’s game" (1, 1) at turn: 1 with reason: 'aborted'.
20220205 17:55:15 debug server: saving replay: 2p__Elensefar_Courtyard_Turn_1_(1).bz2
20220205 17:55:15 info server: <unknown address>        connection closed
20220205 17:55:15 info server: 127.0.0.1        p1      has logged off
20220205 17:55:15 error server: Caught exception from handler: 
Throw location unknown (consider using BOOST_THROW_EXCEPTION)
Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::current_exception_std_exception_wrapper<std::runtime_error> >
std::exception::what: 
boost::system::system_error
20220205 17:55:15 info server: <unknown address>        connection closed
20220205 17:55:15 info server: 127.0.0.1        p2      has logged off
```
and results in lsof showing a `sock` line.

---

Then:

> coro_send_doc(socket, *(queues[socket].front()), yield);, this thing is most definitely not exception safe
> if exception gets thrown from here it won't erase socket from that list then that's how it gets leaked

which dcacc5d3b018f7a85dbf58657d9081467e5e1af7 seems to have fixed in my local testing.

**Comment by Pentarctagon** [comment]
Closing since this *seems* to have been resolved as of 2dcd7ca0fb53fadad42b36ac52d81926a26e215a

**Comment by Pentarctagon** [comment]
Re-pinning this since the file descriptor leak will still cause wesnothd to crash and restart periodically, which isn't a good thing if we want people to play online.

**Comment by Pentarctagon** [comment]
Closed with commits dc7100653e1be31309d8312b511d169b0691ecb6, 88f0fbe18f1c0b57885b2a24ba3ba6b6370aee1e, f13ab34373c98b2bd64352704e76e48da09f5d1f.
